File: src/view_generator.py
"""
Overstand - View Generator

This module handles the generation of HTML-based views like fret position tables.
"""

from typing import Dict, Any
from geometry_engine import calculate_fret_positions
from parameter_registry import InstrumentFamily

def generate_fret_positions_view(params: Dict[str, Any]) -> Dict[str, Any]:
    """Generate fret positions data for display."""
    vsl = params.get('vsl') or 0
    instrument_family = params.get('instrument_family') or InstrumentFamily.VIOLIN.name

    if params.get('no_frets') is not None:
        no_frets = params.get('no_frets')
    elif instrument_family == InstrumentFamily.VIOL.name:
        no_frets = 7
    elif instrument_family == InstrumentFamily.GUITAR_MANDOLIN.name:
        no_frets = 20
    else:
        no_frets = 0

    if no_frets == 0:
        return {'available': False, 'message': 'Fret positions not applicable for violin family'}

    fret_positions = calculate_fret_positions(vsl, no_frets)

    html = '<div class="fret-table-container">'
    html += '<table class="fret-table">'
    html += '<thead><tr><th>Fret</th><th>Distance from Nut (mm)</th><th>Distance from Previous Fret (mm)</th></tr></thead>'
    html += '<tbody>'

    prev_pos = 0
    for i, pos in enumerate(fret_positions):
        fret_num = i + 1
        from_nut = pos
        from_prev = pos - prev_pos
        html += f'<tr><td>{fret_num}</td><td>{from_nut:.1f}</td><td>{from_prev:.1f}</td></tr>'
        prev_pos = pos

    html += '</tbody></table></div>'

    return {
        'available': True,
        'html': html,
        'vsl': vsl,
        'no_frets': no_frets
    }

File: src-ts/view_generator.ts
/**
 * Overstand - View Generator (TypeScript port)
 *
 * Handles the generation of HTML-based views like fret position tables.
 *
 * Ported from src/view_generator.py
 */

import { calculateFretPositions } from './geometry_engine'
import type { Params } from './types'
import { getNumParam, getStringParam } from './utils'

// ============================================================================
// generateFretPositionsView
// ============================================================================

/**
 * Generate fret positions data for display.
 *
 * Returns an object with either:
 * - { available: false, message: string } - frets not applicable
 * - { available: true, html: string, vsl: number, no_frets: number } - fret table HTML
 */
export function generateFretPositionsView(params: Params): {
  available: boolean
  message?: string
  html?: string
  vsl?: number
  no_frets?: number
} {
  const vsl = getNumParam(params, 'vsl')
  const instrument_family = getStringParam(params, 'instrument_family', 'VIOLIN')

  let no_frets: number
  if (params['no_frets'] != null) {
    no_frets = params['no_frets'] as number
  } else if (instrument_family === 'VIOL') {
    no_frets = 7
  } else if (instrument_family === 'GUITAR_MANDOLIN') {
    no_frets = 20
  } else {
    no_frets = 0
  }

  if (no_frets === 0) {
    return {
      available: false,
      message: 'Fret positions not applicable for violin family',
    }
  }

  const fret_positions = calculateFretPositions(vsl, no_frets)

  let html = '<div class="fret-table-container">'
  html += '<table class="fret-table">'
  html +=
    '<thead><tr><th>Fret</th><th>Distance from Nut (mm)</th><th>Distance from Previous Fret (mm)</th></tr></thead>'
  html += '<tbody>'

  let prev_pos = 0
  for (let i = 0; i < fret_positions.length; i++) {
    const pos = fret_positions[i]!
    const fret_num = i + 1
    const from_nut = pos
    const from_prev = pos - prev_pos
    html += `<tr><td>${fret_num}</td><td>${from_nut.toFixed(1)}</td><td>${from_prev.toFixed(1)}</td></tr>`
    prev_pos = pos
  }

  html += '</tbody></table></div>'

  return {
    available: true,
    html,
    vsl,
    no_frets,
  }
}
